rust_src/src/eval.rs

```diff
@@ -1,91 +1,138 @@
+
+    if cond != Qnil {
+        LispObject::from_raw(unsafe { eval_sub(cell.cdr().as_cons_or_error().car().to_raw()) })
+    } else {
+        progn(cell.cdr().as_cons_or_error().cdr())
+    }
 }
 
-macro_rules! message_with_string {
-    ($str:expr, $obj:expr, $should_log:expr) => {
-        unsafe {
-            ::remacs_sys::message_with_string($str.as_ptr() as *const ::libc::c_char,
-                                              $obj.to_raw(),
-                                              $should_log);
+/// Try each clause until one succeeds.
+/// Each clause looks like (CONDITION BODY...).  CONDITION is evaluated
+/// and, if the value is non-nil, this clause succeeds:
+/// then the expressions in BODY are evaluated and the last one's
+/// value is the value of the cond-form.
+/// If a clause has one element, as in (CONDITION), then the cond-form
+/// returns CONDITION's value, if that is non-nil.
+/// If no clause succeeds, cond returns nil.
+/// usage: (cond CLAUSES...)
+#[lisp_fn(min = "0", unevalled = "true")]
+pub fn cond(args: LispObject) -> LispObject {
+    let mut val = Qnil;
+
+    for clause in args.iter_cars_safe() {
+        let cell = clause.as_cons_or_error();
+        val = unsafe { eval_sub(cell.car().to_raw()) };
+        if val != Qnil {
+            let tail = cell.cdr();
+            if tail.is_not_nil() {
+                val = progn(tail).to_raw();
+            }
+            break;
         }
     }
+
+    LispObject::from_raw(val)
 }
 
-/// Macro to format an error message.
-/// Replaces error() in the C layer.
-macro_rules! error {
-    ($str:expr) => {{
-        let strobj = unsafe {
-            ::remacs_sys::make_string($str.as_ptr() as *const ::libc::c_char,
-                                      $str.len() as ::libc::ptrdiff_t)
-        };
-        xsignal!(::remacs_sys::Qerror, $crate::lisp::LispObject::from_raw(strobj));
-    }};
-    ($fmtstr:expr, $($arg:expr),*) => {{
-        let formatted = format!($fmtstr, $($arg),*);
-        let strobj = unsafe {
-            ::remacs_sys::make_string(formatted.as_ptr() as *const ::libc::c_char,
-                                      formatted.len() as ::libc::ptrdiff_t)
-        };
-        xsignal!(::remacs_sys::Qerror, $crate::lisp::LispObject::from_raw(strobj));
-    }};
+/// Eval BODY forms sequentially and return value of last one.
+/// usage: (progn BODY...)
+#[lisp_fn(min = "0", unevalled = "true")]
+pub fn progn(body: LispObject) -> LispObject {
+    body.iter_cars_safe()
+        .map(|form| unsafe { eval_sub(form.to_raw()) })
+        .last()
+        .map_or_else(|| LispObject::constant_nil(), LispObject::from_raw)
 }
 
-/// Macro to format a "wrong argument type" error message.
-macro_rules! wrong_type {
-    ($pred:expr, $arg:expr) => {
-        xsignal!(::remacs_sys::Qwrong_type_argument, LispObject::from_raw($pred), $arg);
-    };
+/// Evaluate BODY sequentially, discarding its value.
+#[no_mangle]
+pub extern "C" fn prog_ignore(body: Lisp_Object) {
```

Comment:
Is this used elsewhere? I can only see one reference to it, so I'm wondering if we need it.

Reply:
It is used twice in eval.c and once in bytecode.c.